product_multi_company/models/product_product.py

```diff
@@ -0,0 +1,49 @@
+# Copyright 2020 Kevin Graveman <k.graveman@onestein.nl>
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+from odoo import api, fields, models
+
+
+class ProductProduct(models.Model):
+    _inherit = ["multi.company.abstract", "product.product"]
+    _name = "product.product"
+    _description = "Product Variant (Multi-Company)"
+
+    company_ids = fields.Many2many(
+        string="Companies",
+        comodel_name="res.company",
+        default=lambda self: self._default_company_ids(),
+    )
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        """Ensure the template initially has the same companies as the variant."""
+        for values in vals_list:
+            if not values.get("company_ids") and values.get("product_tmpl_id"):
+                values["company_ids"] = [
+                    (
+                        6,
+                        0,
+                        self.env["product.template"]
+                        .browse(values.get("product_tmpl_id"))
+                        .company_ids.ids,
+                    )
+                ]
+        res = super(ProductProduct, self).create(vals_list)
+        if "from_create_variants" not in self._context:
+            res.product_tmpl_id.write({"company_ids": [(6, 0, res.company_ids.ids)]})
+        return res
+
+    def write(self, values):
+        """Ensure the template's companies are a superset of the variant's."""
+        if self.env.context.get("company_inverse"):
+            return super(ProductProduct, self).write(values)
+        if values.get("company_ids") and values["company_ids"][0][0] == 6:
+            # Add companies to the template, but don't remove them.
+            companies_to_add = list(
+                set(values.get("company_ids")[0][2]).difference(self.company_ids.ids)
+            )
+            companies_to_add.extend(self.product_tmpl_id.company_ids.ids)
+            if companies_to_add:
+                self.product_tmpl_id.write({"company_ids": [(6, 0, companies_to_add)]})
+        return super(ProductProduct, self).write(values)
```

Comment:
All this code doesn't exist in the previous branch. Actually, there was no `product_product.py` file. Only a small `product_template.py`:

https://github.com/OCA/multi-company/tree/ae4570d4f327ce5e0ad3779bbcde3369b82c3acb/product_multi_company/models

I observe also that in the commit 9633be1b29f0a04eba9fa5fa3fc0efd7853f006e, labeled `[MIG] product_multi_company: black, isort, prettier`, you're not actually just passing the automated pre-commit reformats. You're refactoring the module completely.

Most of the logic you add there should be unnecessary at first sight, as it should exist already in the inherited `multi.company.abstract` model.

I'm confused. Why did you have to refactor the module? Didn't the abstract model work? Why did you hide these big changes in the reformat commit?

Reply:
@yajo 

In v13, they added those extra changes to avoid the below error.
https://github.com/OCA/multi-company/issues/252
And it is still happening in the latest v15 code.

Detailed info regarding the issues is in the below link
https://github.com/OCA/multi-company/pull/264 

We are having the same issue in migrated code too.

I think we need to add those back. 

Thanks

Reply:
@yajo 
Creating a related stored company_ids field on the product.product solves this issue.
Shall we go with this instead of having those all changes from v13?
I can't think of any disadvantages with this method.